### PR TITLE
feat(FN-1726): add fee to existing payment button

### PR DESCRIPTION
**Pull request by sw-mattw**
## Introduction :pencil2:
Renders an "Add reported fee(s) to an existing payment" button on the add payment screen. The button appears only when there are eligible existing payments for the selected fee record(s).

## Resolution :heavy_check_mark:
- Added a method to the payment repository to check for the existence of a payment entity based on report ID and currency.
- Added a new flag in the selected fee records details API response to indicate whether adding the selected fees to an existing payment is possible. This functionality considers the statuses of the selected fee records (must all be in `TO_DO`) and the existence of payments with both a matching payment currency and at least one fee record with the status `DOES_NOT_MATCH`.
- Updated the add payment UI template to display the new button for adding fee records to an existing payment when applicable. The button's visibility is controlled by a flag passed in by the view model.

Eligible existing payments found, single fee:
UKEF_FN-1726_Single-Fee

Eligible existing payments found, multiple fee:
UKEF_FN-1726_Multiple-Fees

No eligible existing payments found:
UKEF_FN-1726_No-eligible-existing-payments

POST request made after clicking the button (included for payload visibility):
POST request result

## Miscellaneous :heavy_plus_sign:
N/A

### CODE REVIEW

**Review by RLavender98 — DISMISSED** [review]
Looks good, mostly just a few comments around naming!

File: dtfs-central-api/api-tests/v1/utilisation-reports/get-selected-fee-record-details.api-test.ts [hunk 1]
```diff
@@ -93,7 +112,46 @@ describe('GET /v1/utilisation-reports/:id/selected-fee-records-details', () => {
           },
         ],
         payments: [],
+        canAddToExistingPayment: true,
+      });
+    });
+
+    it("gets selected fee record details with canAddToExistingPayment set to false when a payment exists in the same currency but has no fee records with the 'does not match' status", async () => {
+      // Act
+      const response: CustomResponse = await testApi.get(getUrl(reportId), { feeRecordIds: [47] });
```
[the remark this answers is not in the document]
Comment (sw-mattw) [reply dtfs-central-api/api-tests/v1/utilisation-reports/get-selected-fee-record-details.api-test.ts:132]: Yep that sounds sensible, changes in c4304b6b093cf4f0171c8565cdf154a99d0391ce

File: dtfs-central-api/src/v1/controllers/utilisation-report-service/get-selected-fee-records-details.controller/helpers.test.ts [hunk 2]
```diff
@@ -168,14 +170,76 @@ describe('get selected fee record details controller helpers', () => {
...
+      // Assert
+      expect(result).toEqual(false);
+    });
   });
 
-  function aFeeRecord() {
-    return FeeRecordEntityMockBuilder.forReport(UtilisationReportEntityMockBuilder.forStatus('RECONCILIATION_IN_PROGRESS').build()).build();
+  function aFeeRecord(): FeeRecordEntity {
+    const aUtilisationReport = UtilisationReportEntityMockBuilder.forStatus('RECONCILIATION_IN_PROGRESS').build();
+    return FeeRecordEntityMockBuilder.forReport(aUtilisationReport).withPaymentCurrency('GBP').withStatus('TO_DO').build();
```
[the remark this answers is not in the document]
Comment (sw-mattw) [reply dtfs-central-api/src/v1/controllers/utilisation-report-service/get-selected-fee-records-details.controller/helpers.test.ts:245]: Renamed function - 6c6c067e32f3d567be620b55e43af3997fee9e94

File: trade-finance-manager-ui/component-tests/utilisation-reports/add-payment.component-test.js [hunk 3]
```diff
@@ -348,4 +348,58 @@ describe(page, () => {
...
+    const addPaymentViewModel = {
+      ...anAddPaymentViewModel(),
+      canAddToExistingPayment: true,
+    };
+    const wrapper = render(addPaymentViewModel);
+
+    // Assert
+    const addFeesToExistingPaymentButtonSelector = '[data-cy="add-fees-to-an-existing-payment-button"]';
+    wrapper.expectElement(addFeesToExistingPaymentButtonSelector).toExist();
+    wrapper.expectElement(addFeesToExistingPaymentButtonSelector).toHaveAttribute('value', 'Add reported fee to an existing payment');
```
[the remark this answers is not in the document]
Comment (sw-mattw) [reply trade-finance-manager-ui/component-tests/utilisation-reports/add-payment.component-test.js:364]: Added href and class assertions for this button thanks - 948fa846122987005c04edb10aba38e3f533963c

File: trade-finance-manager-ui/component-tests/utilisation-reports/add-payment.component-test.js [hunk 4]
```diff
@@ -348,4 +348,58 @@ describe(page, () => {
 
     wrapper.expectElement(`[data-cy="remove-selected-fees-button"]`).notToExist();
   });
+
+  it('should display singular "Add reported fee to an existing payment" button when hasExistingPayments is true and only one payment exists', () => {
```
[the remark this answers is not in the document]
Comment (sw-mattw) [reply trade-finance-manager-ui/component-tests/utilisation-reports/add-payment.component-test.js]: Ah yes, thank you, missed this. Resolved in 6c6c067e32f3d567be620b55e43af3997fee9e94

File: dtfs-central-api/src/repositories/payment-repo/payment.repo.ts [hunk 5]
```diff
@@ -23,4 +24,24 @@ export const PaymentRepo = SqlDbDataSource.getRepository(PaymentEntity).extend({
...
+
+  /**
+   * Checks if a payment entity exists for a report with the supplied id,
+   * matching the supplied currency, and having at least one fee record
+   * with status 'DOES_NOT_MATCH'
+   * @param reportId - The report id
+   * @param currency - The payment currency to search by
+   * @returns True if a matching payment entity exists, false otherwise
+   */
+  async existsByReportIdCurrencyWithUnmatchedFeeRecords(reportId: number, currency: Currency): Promise<boolean> {
```
[the remark this answers is not in the document]
Comment (sw-mattw) [reply dtfs-central-api/src/repositories/payment-repo/payment.repo.ts]: Renamed to `existsUnmatchedPaymentOfCurrencyForReportWithId` as I think that sounds most suitable - 6c6c067e32f3d567be620b55e43af3997fee9e94

File: dtfs-central-api/src/v1/controllers/utilisation-report-service/get-selected-fee-records-details.controller/helpers.test.ts [hunk 6]
```diff
@@ -168,14 +170,76 @@ describe('get selected fee record details controller helpers', () => {
...
+      // Act
+      const result = await mapToSelectedFeeRecordDetails(bankId, aReportPeriod(), [aFeeRecord()], canAddToExistingPayment);
+
+      // Assert
+      expect(result.canAddToExistingPayment).toEqual(canAddToExistingPayment);
+    });
+  });
+
+  describe('canFeeRecordsBeAddedToExistingPayment', () => {
+    it('returns true when payment exists on report with matching reported currency', async () => {
```
[the remark this answers is not in the document]
Comment (sw-mattw) [reply dtfs-central-api/src/v1/controllers/utilisation-report-service/get-selected-fee-records-details.controller/helpers.test.ts]: Resolved in 6c6c067e32f3d567be620b55e43af3997fee9e94

File: dtfs-central-api/src/v1/controllers/utilisation-report-service/get-selected-fee-records-details.controller/helpers.test.ts [hunk 7]
```diff
@@ -168,14 +170,76 @@ describe('get selected fee record details controller helpers', () => {
...
+      // Arrange
+      jest.spyOn(PaymentRepo, 'existsByReportIdCurrencyWithUnmatchedFeeRecords').mockResolvedValue(true);
```
[the remark this answers is not in the document]
Comment (sw-mattw) [reply dtfs-central-api/src/v1/controllers/utilisation-report-service/get-selected-fee-records-details.controller/helpers.test.ts]: Ah yes, added spy expectations to this test and some of the others in this file thanks - 0063385a92f3f130362fba2e06234793514e3290

File: dtfs-central-api/src/v1/controllers/utilisation-report-service/get-selected-fee-records-details.controller/helpers.test.ts [hunk 8]
```diff
@@ -168,14 +170,76 @@ describe('get selected fee record details controller helpers', () => {
...
+      const feeRecord = FeeRecordEntityMockBuilder.forReport(aUtilisationReport).withPaymentCurrency('GBP').withStatus('READY_TO_KEY').build();
+
+      // Act
+      const result = await canFeeRecordsBeAddedToExistingPayment('123', [feeRecord]);
+
+      // Assert
+      expect(result).toEqual(false);
+    });
+
+    it('returns false when no feerecords', async () => {
```
[the remark this answers is not in the document]
Comment (sw-mattw) [reply dtfs-central-api/src/v1/controllers/utilisation-report-service/get-selected-fee-records-details.controller/helpers.test.ts]: Added space in 6c6c067e32f3d567be620b55e43af3997fee9e94

File: dtfs-central-api/src/v1/controllers/utilisation-report-service/get-selected-fee-records-details.controller/helpers.ts [hunk 9]
```diff
@@ -31,10 +32,27 @@ const getTotalReportedPayments = (feeRecords: SelectedFeeRecordDetails[]): Curre
   amount: feeRecords.map((feeRecord) => feeRecord.reportedPayments.amount).reduce((total, currentAmount) => total + currentAmount, 0),
 });
 
+export const canFeeRecordsBeAddedToExistingPayment = async (reportId: string, feeRecords: FeeRecordEntity[]): Promise<boolean> => {
+  if (feeRecords.length === 0) {
+    return false;
+  }
+
+  const allFeeRecordsHaveStatusToDo = feeRecords.every((record) => record.status === 'TO_DO');
```
[the remark this answers is not in the document]
Comment (sw-mattw) [reply dtfs-central-api/src/v1/controllers/utilisation-report-service/get-selected-fee-records-details.controller/helpers.ts]: Ah that's a nice spot, an early exit check is nice - added in 5ca8383bcf67afc2077ca16eb6caf2b729eec981

File: dtfs-central-api/api-tests/v1/utilisation-reports/get-selected-fee-record-details.api-test.ts [hunk 10]
```diff
@@ -25,23 +33,34 @@ describe('GET /v1/utilisation-reports/:id/selected-fee-records-details', () => {
...
     .withExporter('Test company')
     .withFeesPaidToUkefForThePeriod(100)
     .withFeesPaidToUkefForThePeriodCurrency('GBP')
     .withPaymentCurrency('GBP')
+    .withPayments([aPaymentInGBP])
+    .withStatus('DOES_NOT_MATCH')
     .build();
-  const anotherFeeRecord = FeeRecordEntityMockBuilder.forReport(utilisationReport)
-    .withId(46)
+  const aFeeRecordInUSD = aFeeRecordWithPaymentCurrency(47, 'USD');
```
[the remark this answers is not in the document]
Comment (sw-mattw) [reply dtfs-central-api/api-tests/v1/utilisation-reports/get-selected-fee-record-details.api-test.ts]: Added StatusToDo reference to this function in 6c6c067e32f3d567be620b55e43af3997fee9e94 and 471b0b5f0cf293a43c63870e19d59015cb17215d

**Review by RLavender98 — APPROVED** [review]
LGTM!

**Pull request merged**